File: lib/physics/body.go
package physics

import (
	"errors"

	"buildblast/lib/vmath"
)

type Body struct {
	Pos          vmath.Vec3
	Vel          vmath.Vec3
	Dir          vmath.Vec3
	HalfExtents  vmath.Vec3
	CenterOffset vmath.Vec3
}

func (b *Body) Box() *Box {
	return NewBoxOffset(b.Pos, b.HalfExtents, b.CenterOffset)
}

func CloneBody(body *Body) *Body {
	return &Body{
		body.Pos, body.Vel, body.Dir, body.HalfExtents, body.CenterOffset,
	}
}

// Protocol stuff

func (b *Body) ToProto() []byte {
	buf := make([]byte, 5*3*8)
	buf = append(buf, b.Pos.ToProto())
	buf = append(buf, b.Vel.ToProto())
	buf = append(buf, b.Dir.ToProto())
	buf = append(buf, b.HalfExtents.ToProto())
	buf = append(buf, b.CenterOffset.ToProto())
	return buf
}

func (b *Body) FromProto(buf []byte) (int, error) {
	if len(buf) < 5*3*8 {
		return 0, errors.New("Buffer too small: Body")
	}
	size := 3 * 8
	b.Pos.FromProto(buf[0*size:])
	b.Vel.FromProto(buf[1*size:])
	b.Dir.FromProto(buf[2*size:])
	b.HalfExtents.FromProto(buf[3*size:])
	b.CenterOffset.FromProto(buf[4*size:])
	return 5 * 3 * 8, nil
}

File: client/js/physics/body.js
define(function(require){

var THREE = require("THREE");
var Box = require("./box");

return function Body(pos, vel, dir, halfExtents, centerOffset) {
	var self = this;

	self.pos = pos || new THREE.Vector3(0, 0, 0);
	self.vel = vel || new THREE.Vector3(0, 0, 0);
	self.dir = dir || new THREE.Vector3(0, 0, 0);
	self.halfExtents = halfExtents || new THREE.Vector3(0, 0, 0);
	self.centerOffset = centerOffset || new THREE.Vector3(0, 0, 0);
	self.box = new Box(self.halfExtents, self.centerOffset);

	self.clone = function() {
		return new Body(self.pos.clone(), self.vel.clone(), self.dir.clone(),
				self.halfExtents.clone(), self.centerOffset.clone());
	};

	self.lerp = function(other, frac) {
		self.pos.lerp(other.pos, frac);
		self.vel.lerp(other.vel, frac);
		self.dir.lerp(other.dir, frac);
		self.halfExtents.lerp(other.halfExtents, frac);
		self.centerOffset.lerp(other.centerOffset, frac);
	}
}

});